### ISSUE TITLE
Can it build on Linux to windows

### ISSUE THREAD

**Issue by TomatoMr** [rust-pcap/pcap#246]
I want to build on Linux to Windows, can it supports?

**Comment by Wojtek242** [comment]
Yes, please see the README for instructions

**Comment by Wojtek242** [comment]
Oh sorry, misunderstood your question. I thought you asked if it builds on Windows.

Building from Linux to Windows should work since we support Windows builds, but we don't have it documented.

**Comment by TomatoMr** [comment]
> 

😁Thank you very much for your reply. I would be very grateful if there are documents or construction guidelines.

**Comment by TomatoMr** [comment]
I used the build tool: [cargo-xwin ](https://github.com/messense/cargo-xwin) , and tried to run:
```shell
cargo xwin build --target x86_64-pc-windows-msvc
```
then, I got the failed info:
```shell
  = note: lld-link: error: could not open 'wpcap.lib': No such file or directory
          lld-link: error: could not open 'pcap.lib': No such file or directory
```
![image](https://user-images.githubusercontent.com/15191500/184536588-4ec9a6fd-d198-43bf-825f-a2bc598cef97.png)

I have downloaded the SDK of WinPcap and copied wpcap.lib and packet.lib in the x64 directory to Linux. But I can't find pcap.lib

**Comment by Wojtek242** [comment]
I'm no expert on cross-compiling so I can't help you here. However, I'll reopen the issue so that somebody else perhaps will see this and answer.

**Comment by TomatoMr** [comment]
> 

tks🌹

**Comment by Stargateur** [comment]
have you try use `LIB` env variable to pin point the location of `wpcap.lib` ?

**Comment by TomatoMr** [comment]
> have you try use `LIB` env variable to pin point the location of `wpcap.lib` ? also try to just put the file at the root of the project. Should work.

Yes, I tried it, it didn't work. I also tried to copy and rename `wpcap.lib` to `pcap.lib`, but it still failed with the error: `lld-link: error: could not open 'pcap.lib': No such file or directory`. I can't find `pcap.lib`.

**Comment by robs-zeynet** [comment]
You might try building with a build.rs (which puts the build information into the build process) instead of the manual environmental variable.  The following build.rs works for me on a "built for windows, on windows" build which I haven't tried from Linux, but should be a reasonable starting point.  It assumes the same directory structure as what comes out of the npcap dev libs, e.g., that if you want a x64 bit build, you want to use the "lib/x64" subdir.
```
robsh@LAPTOP-O7RS71IJ MINGW64 ~/git/go3x (main)
$ find lib/
lib/
lib/x64
lib/x64/Packet.lib
lib/x64/wpcap.lib

robsh@LAPTOP-O7RS71IJ MINGW64 ~/git/go3x (main)
$ ls build.rs
build.rs

$ cat build.rs
use std::env;
use std::path::Path;

fn main() {
    let dir = env::var("CARGO_MANIFEST_DIR").unwrap();
    println!(
        "cargo:rustc-link-search=native={}",
        Path::new(&dir).join("lib/x64").display()
    );
}
```

**Comment by TomatoMr** [comment]
> You might try building with a build.rs (which puts the build information into the build process) instead of the manual environmental variable. The following build.rs works for me on a "built for windows, on windows" build which I haven't tried from Linux, but should be a reasonable starting point. It assumes the same directory structure as what comes out of the npcap dev libs, e.g., that if you want a x64 bit build, you want to use the "lib/x64" subdir.
> 
> ```
> robsh@LAPTOP-O7RS71IJ MINGW64 ~/git/myproj (main)
> $ find lib/
> lib/
> lib/x64
> lib/x64/Packet.lib
> lib/x64/wpcap.lib
> 
> robsh@LAPTOP-O7RS71IJ MINGW64 ~/git/myproj (main)
> $ ls build.rs
> build.rs
> 
> $ cat build.rs
> use std::env;
> use std::path::Path;
> 
> fn main() {
>     let dir = env::var("CARGO_MANIFEST_DIR").unwrap();
>     println!(
>         "cargo:rustc-link-search=native={}",
>         Path::new(&dir).join("lib/x64").display()
>     );
> }
> ```

@robs-zeynet thank you, i had tried it, but it not worked

**Comment by jonas-kell** [comment]
Ok, I mean this is years old so I do not know if this is relevant to anyone. But I just needed this and figured it out:

```cmd
sudo apt install mingw-w64-tools binutils-mingw-w64
mkdir generated_libs
cd generated_libs
curl https://npcap.com/dist/npcap-1.87.exe > lib.exe
7z x lib.exe  # Always

gendef wpcap_x64.dll
mv wpcap_x64.def wpcap.def
x86_64-w64-mingw32-dlltool -d wpcap.def -l libwpcap.a

gendef Packet_x64.dll
mv Packet_x64.def Packet.def
x86_64-w64-mingw32-dlltool -d Packet.def -l libpacket.a
``` 

This will generate a folder that includes vendored build dependencies for a specific pcap version.

Then in the folder `.cargo` you need a file `config.toml`:
```toml
[target.x86_64-pc-windows-gnu]
rustflags = [
    "-L", "/app/generated_libs",
    "-l", "wpcap",
    "-l", "packet"
]
```

(the path is app, because that is where my project is mapped into the docker container).

The docker compose file that performs the cross compilation is 
```yml
services:
  build-windows:
    build:
      context: ./
      dockerfile: ./docker/compile_windows/Dockerfile
    volumes:
      - rust_build_cache_windows:/usr/local/cargo/registry/
      - ./:/app/
    working_dir: /app
    environment:
      - CARGO_TARGET_DIR=win_target
    command: cargo build --release --target x86_64-pc-windows-gnu

volumes:
  rust_build_cache_windows:
```

With the `docker/compile_windows/Dockerfile`:
```yml
FROM rust:1.94

RUN apt-get update && apt-get install -y \
    libx11-dev \
    libxcursor-dev \
    libxrandr-dev \
    libxinerama-dev \
    libxi-dev \
    libgl1-mesa-dev \
    libwayland-dev \
    libxkbcommon0 \
    libxkbcommon-x11-0 \
    libasound2-dev \
    libpulse-dev \
    libudev-dev \
    libpcap-dev \
    pkg-config \
    && rm -rf /var/lib/apt/lists/*

RUN apt-get update && apt-get install -y mingw-w64 \
    && rm -rf /var/lib/apt/lists/*

RUN rustup target add x86_64-pc-windows-gnu

WORKDIR /app/
```
(I think most dependencies can be omitted in general, but my project requires all of them, so I will not test now which are relevant. I suppose you can omit all from `libx11-dev` to `libpulse-dev`, but not tested).
